docs/design/speed-test.md

```diff
@@ -0,0 +1,210 @@
+role in disaster recovery strategies, it is essential to have a reliable way to test upload speeds.
+
+This Upload Speed Test CRD will help in such scenarios by proactively measuring the upload speed from the OpenShift cluster to the cloud storage configured in OADP’s BSL. By identifying
+performance bottlenecks ahead of backup or restore operations, administrators can take corrective actions to avoid failures during critical operations.
+
+## Goals
+
+-   Measure upload speeds for a given BackupStorageLocation (BSL) config settings and credentials
+-   Measure upload speeds from the OpenShift cluster to the configured object storage.
+-   Update the test results in the status of the UploadSpeedTest CRD for visibility and tracking.
+-   Enable smooth integration with OADP operator-managed backup storage location configurations.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- UploadSpeedTest (UST) CRD:
+  - Captures the BSL configuration, test file size and test timeout used for testing directly in the CRD.
+  - Holds the test results, including success/failure and upload speed.
+- UploadSpeedTest (UST) Controller:
+  - Watches for changes to the UploadSpeedTest CRs and executes uploads based on the provided BSL configuration. 
+  - Uses appropriate cloud SDKs (AWS, Azure, or GCP) to perform uploads.
+  - Updates the status of UST CR with upload speed, upload success/failure and error message if any.
+- OADP/DPA-UST Integration:
+  - Fetches BackupStorageLocation (BSL) details from the OADP operator’s DPA CR. 
+  - Populates the UploadSpeedTest CR with relevant configurations from the BSL.
+  - Creates the UST CR for the BSL configured in DPA instance.
+## Detailed Design
+
+Proposed Upload Speed Test CRD:
+
+The UploadSpeedTest CRD structure includes fields to capture backup storage location configurations, upload test parameters, and the location of cloud provider secrets.
+
+```yaml
+apiVersion: oadp.openshift.io/v1alpha1
+kind: UploadSpeedTest
+metadata:
+  name: my-upload-speed-test
+spec:
+  backupLocation:
+    velero:
+      provider: aws # Cloud provider type (aws, azure, gcp)
+      default: true
+      objectStorage:
+        bucket: sample-bucket
+        prefix: velero
+      config:
+        region: us-east-1
+        profile: "default"
+        insecureSkipTLSVerify: "true"
+      credential:
+        name: cloud-credentials  # Secret containing cloud credentials
+        key: cloud
+  uploadSpeedTestConfig:
+    fileSize: "100MB"    # File size for the test upload
+    testTimeout: "60s"   # Maximum duration allowed for the test
+  cloudProviderSecretRef:
+    name: cloud-credentials
+    namespace: openshift-adp
+```
```

Comment:
Could we change here to only have 2 fields in spec (`uploadSpeedTestConfig` and `backupLocationName`)?

`cloudProviderSecretRef` can be fetched from BSL, right? So does not seem needed

Having BSL spec copied here makes it harder for user to manually create one (even thinking about DPA integration, if BSL spec changes, UST must change. Having just name reference would not simplify workflow?)

Reply:
Done, removed `cloudProviderSecretRef` from UST CRD spec, tested AWS POC as well.